core/safeclient/client.go

```diff
@@ -0,0 +1,930 @@
+		c.createClient = createClient
+	}
+}
+
+func (c *SafeEthClient) reinit() bool {
+	c.clientLock.Lock()
+	defer c.clientLock.Unlock()
+
+	client, err := c.createClient(c.rpcUrl, c.logger)
+	if err != nil {
+		c.logger.Error("Failed to reinitialize client", "err", err)
+		return false
+	}
+
+	c.Client = client
+	c.logger.Info("Successfully reinitialized client")
+	return true
+}
+
+func (c *SafeEthClient) notifySubscribers(success bool) {
+	c.reinitSubscribersLock.Lock()
+	defer c.reinitSubscribersLock.Unlock()
+
+	for _, ch := range c.reinitSubscribers {
+		ch <- success
+		close(ch)
+	}
+	c.reinitSubscribers = nil
+	c.logger.Debug("Notified subscribers of reinit", "success", success)
+}
+
+func (c *SafeEthClient) handleReinitEvent() {
+	success := c.reinit()
+	c.notifySubscribers(success)
+	if success {
+		return
+	}
+
+	defer c.wg.Done()
+
+	c.isReinitializing = true
+
+	c.wg.Add(1)
+	go func() {
+		defer func() {
+			c.isReinitializing = false
+			c.wg.Done()
+		}()
+
+		reinitTicker := time.NewTicker(c.reinitInterval)
+		defer reinitTicker.Stop()
+
+		for {
+			select {
+			case <-reinitTicker.C:
+				c.logger.Debug("Reinit ticker fired")
+
+				success := c.reinit()
+				c.notifySubscribers(success)
+
+				if success {
+					return
+				}
+			case <-c.closeC:
+				c.logger.Info("Received close signal, stopping reinit handler")
+				return
+			}
+		}
+	}()
+}
+
+func (c *SafeEthClient) handleReinit() {
+	defer c.wg.Done()
+
+	c.isReinitializing = false
+
+	for {
+		select {
+		case <-c.closeC:
+			c.logger.Info("Received close signal, stopping reinit handler")
+			return
+		case <-c.reinitC:
+			if c.isReinitializing {
+				c.logger.Debug("Already reinitializing, ignoring reinit signal")
+				continue
+			}
+
+			c.logger.Debug("Received reinit signal")
+			c.handleReinitEvent()
+		}
+	}
+}
+
+func (c *SafeEthClient) WatchReinit() <-chan bool {
+	ch := make(chan bool, 1)
+	c.reinitSubscribersLock.Lock()
+	c.reinitSubscribers = append(c.reinitSubscribers, ch)
+	c.reinitSubscribersLock.Unlock()
+	c.logger.Debug("Added reinit watcher")
+	return ch
+}
+
+type SafeSubscription struct {
+	sub          ethereum.Subscription
+	lock         sync.Mutex
+	errC         chan error
+	unsubscribed bool
+}
+
+func NewSafeSubscription(sub ethereum.Subscription) *SafeSubscription {
+	return &SafeSubscription{
+		sub:  sub,
+		errC: make(chan error, 1),
+	}
+}
+
+func (s *SafeSubscription) Err() <-chan error {
+	return s.errC
+}
+
+func (s *SafeSubscription) Unsubscribe() {
+	s.lock.Lock()
+
+	if s.unsubscribed {
+		s.lock.Unlock()
+		return
+	}
+
+	s.sub.Unsubscribe()
+	s.unsubscribed = true
+	s.lock.Unlock()
+
+	<-s.errC
+}
+
+func (s *SafeSubscription) SetUnderlyingSub(sub ethereum.Subscription) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.sub.Unsubscribe()
+	s.sub = sub
+}
+
+func (c *SafeEthClient) SubscribeFilterLogs(ctx context.Context, q ethereum.FilterQuery, ch chan<- types.Log) (ethereum.Subscription, error) {
+	c.clientLock.RLock()
```

Comment:
Are locks used for `c.Client.SubscribeFilterLogs` here? If so locks also needed `SubscribeNewHead`

Reply:
That's right, done in [ec52183](https://github.com/NethermindEth/near-sffl/pull/154/commits/ec52183409e00bbd0a749f990582d4a372b88a1c)